pkg/tfshim/sdk-v2/upgrade_state.go

```diff
@@ -32,56 +34,53 @@ func upgradeResourceState(ctx context.Context, p *schema.Provider, res *schema.R
 		if err != nil {
 			return nil, err
 		}
-		version, hasVersion = int(v), true
+		version, hasVersion = v, true
 	}
 
-	// First, build a JSON state from the InstanceState.
-	json, version, err := schema.UpgradeFlatmapState(ctx, version, m, res, p.Meta())
-	if err != nil {
-		return nil, err
-	}
+	// Now, we perform the UpgradeResourceState operation by re-implementing TF's UpgradeResourceState.
 
-	// Next, migrate the JSON state up to the current version.
-	json, err = schema.UpgradeJSONState(ctx, version, json, res, p.Meta())
+	resp, err := schema.NewGRPCProviderServer(p).
+		UpgradeResourceState(ctx, &tfprotov5.UpgradeResourceStateRequest{
+			TypeName: typeName,
+			Version:  version,
+			RawState: &tfprotov5.RawState{Flatmap: m},
```

Comment:
I believe you said the flatmap representation is problematic, right?

Reply:
Under the hood this calls UpgradeFlatmapState

Reply:
It is, and that is fixed in [c9ea19c](https://github.com/pulumi/pulumi-terraform-bridge/pull/1735/commits/c9ea19c2fd466d7f49c98724de0f7c5dd7ee2f50).